### PR TITLE
Add an optional Device parameter to pin_memory/is_pinned that does nothing

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #60547 Make pin_memory/is_pinned use BackendSelect
* **#60201 Add an optional Device parameter to pin_memory/is_pinned that does nothing**
* #59655 Remove some unnecessary functions from CUDAHooks

This is to flush out BC/FC problems with adding this parameter.  Later
PR will actually add the desired functionality.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D29331880](https://our.internmc.facebook.com/intern/diff/D29331880)

### CODE REVIEW

**Comment by ezyang** [comment]
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29331880).

**Review by ngimel — APPROVED** [review]

**Comment by ezyang** [comment]
ah, hard coded jit schema, my favorite

**Comment by ezyang** [comment]
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29331880).

File: aten/src/ATen/native/Memory.cpp [hunk 1]
```diff
@@ -9,14 +9,16 @@
...
   return detail::getCUDAHooks().isPinnedPtr(self.storage().data());
 }
 
-Tensor pin_memory(const Tensor& self) {
+Tensor pin_memory(const Tensor& self, c10::optional<Device> device) {
   if (!self.device().is_cpu()) {
     AT_ERROR("cannot pin '", self.toString(), "' only dense CPU tensors can be pinned");
   }
+  TORCH_CHECK(!device.has_value() || device->is_cuda(), "non-cuda device doesn't have a concept of pinned memory");
   if (self.is_pinned()) {
```
Comment (bdhirsh) [inline aten/src/ATen/native/Memory.cpp:22]: should we be passing `device` into `self.is_pinned()`?

**Review by bdhirsh — APPROVED** [review]

File: aten/src/ATen/native/Memory.cpp (hunk 1 above)
Comment (bdhirsh) [reply aten/src/ATen/native/Memory.cpp:22]: oh i see you delete this in the next PR and let BackendSelect deal with dispatch, I'll keep reading

**Comment by ezyang** [comment]
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29331880).

**Comment by ezyang** [comment]
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29331880).

**Pull request closed without merging**